Move ACT_READ from activity_handler to activity_actor

Saicchi opened the pull request:
#### Summary
SUMMARY: Infrastructure "Move ACT_READ to activity_actor"

#### Purpose of change
Moving an activity from the legacy `activity_handler` to the new `activity_actor` to make adding new features easier

#### Describe the solution
Deleting the read activity handler
Rewriting `avatar::read` and `npc::read` functions and for the newly created `activity_actor`.
Changing tests that used `avatar::do_read` for book indentification to use `avatar::identify` instead.
Add a test for reading books to make sure player gets xp when reading a book they can learn from

#### Describe alternatives you've considered
None

#### Testing
These tests were done by comparing my branch and the branch it diverged from in the exact same save

- **Player reading book**
  - [x] Duration per chapter is the same
  - [x] Duration per player level is the same
  - ~~Duration per NPC level is the same~~ Not possible, read additional context
- **NPC reading book**
  - ~~Duration per level is the same~~ Not possible, read additional context
- **Player/NPC reading book to Player/NPCs**
  - [x] Duration per player level is the same
  - ~~Duration per NPC level is the same~~ Not possible, read additional context 
- **Player/NPC reading to player/npc martial arts book**
  - NPCs can't read martial arts books yet
  - Unchanged
- **Player reading spell books**
  - NPCs can't read spellbooks to the player
  - Unchanged
- **Guidebook ( The Hitchhiker's Guide to the Cataclysm )**
  - Unchanged

#### Additional context

Switched the order of the first two options in the read menu so they feel more natural
**OLD**
![Old Menu](https://user-images.githubusercontent.com/47158232/103109049-ba675300-4625-11eb-87f6-02c775e00d34.png) 
**NEW**
![New Menu](https://user-images.githubusercontent.com/47158232/103109070-bb988000-4625-11eb-8f86-9c6d4961d6f0.png)

Fixed a bug where experience would be adjusted to the player focus and traits instead of the NPC focus and traits
`src/avatar.cpp#L759`
Also, unless I'm mistaken, `focus_pool` for NPCs is not saved, which I think it means that NPCs focus is
reset to 100 everytime the game is loaded

I'd really appreciate if before merging someone compiled this branch and tested it for themselves. Though I've tested it multiple times and did not find any problems nor do I think there are any, I'm scared of it breaking something important like reading books.

actual-nh commented on src/activity_actor_definitions.h:
```diff
@@ -875,5 +879,56 @@ class move_furniture_activity_actor : public activity_actor
+        }
+
+        static void read_book( Character &learner, const cata::value_ptr<islot_book> &islotbook,
+                               SkillLevel &skill_level, const double penalty );
```
```suggestion
                               SkillLevel &skill_level, double penalty );
```

Saicchi replied:
I don't modify `penalty` so I don't see why it shouldn't be a const.

Qrox replied:
`penalty` is passed by value, so whether it is const in the function body does not matter to the outside caller. The error messages says exactly this (`const-qualification of parameters only has an effect in function definitions`).

Saicchi replied:
Oh. I didn't even see the error message, sorry. Clang is really weird.

Qrox commented on src/activity_actor_definitions.h:
```diff
@@ -875,5 +879,56 @@ class move_furniture_activity_actor : public activity_actor
+    private:
+        int moves_total;
+        item_location book;
+        // casting is necessary for serialization
+        int bktype = static_cast<int>( book_type::normal );
```
This doesn't seem to need to be serialized.

Qrox commented on src/activity_actor.cpp:
```diff
@@ -2532,6 +2538,466 @@ std::unique_ptr<activity_actor> milk_activity_actor::deserialize( JsonIn &jsin )
+        return;
+    }
+
+    if( book->type->use_methods.count( "MA_MANUAL" ) ) {
+        bktype = static_cast<int>( book_type::martial_art );
+    }
```
Instead it can always be retrieved from the book type.

Saicchi replied:
I use it mainly for the code in `read_activity_actor::do_turn`
```c++
if( bktype == static_cast<int>( book_type::martial_art ) && one_in( 3 ) ) {
    who.mod_stamina( -1 );
}
```
that way I won't need to call `book->type->use_methods.count( "MA_MANUAL" )` every turn and only once at the start. I could check if its a martial arts book when deserializing but I would like to keep that function simple like the other deserialization functions. I also used a `enum` for book type instead of a `bool` to accommodate future different book types.

Qrox replied:
You can instead cache the value when `do_turn` is called (use `cata::optional`).

```
cata::optional<book_type> bktype;
```

```
if( !bktype.has_value() ) {
    bktype = book->type->use_methods.count( "MA_MANUAL" );
}
```

Saicchi replied:
Thanks. I didn't know how to cache.

actual-nh commented on src/avatar.cpp:
```diff
@@ -459,183 +457,180 @@ bool avatar::read( item &it, const bool continuous )
+
+        if( !learners.empty() ) {
+            add_header( _( "Read until this NPC gains a level:" ) );
+            for( const std::pair<npc *, std::string> &elem : learners ) {
```
```suggestion
            for( const std::pair<npc *const, std::string> &elem : learners ) {
```
Not sure if the first `const` needs removing.

actual-nh commented on src/avatar.cpp:
```diff
@@ -459,183 +457,180 @@ bool avatar::read( item &it, const bool continuous )
-                return false;
+        if( !fun_learners.empty() ) {
+            add_header( _( "Reading for fun:" ) );
+            for( const std::pair<npc *, std::string> &elem : fun_learners ) {
```
```suggestion
            for( const std::pair<npc *const, std::string> &elem : fun_learners ) {
```
Ditto.

actual-nh commented on src/avatar.cpp:
```diff
@@ -459,183 +457,180 @@ bool avatar::read( item &it, const bool continuous )
-                return false;
+        if( !nonlearners.empty() ) {
+            add_header( _( "Not participating:" ) );
+            for( const std::pair<npc *, std::string> &elem : nonlearners ) {
```
```suggestion
            for( const std::pair<npc *const, std::string> &elem : nonlearners ) {
```
Ditto.

actual-nh commented on src/avatar.cpp:
```diff
@@ -459,183 +457,180 @@ bool avatar::read( item &it, const bool continuous )
+
+    // Don't include the reader as it would be too redundant.
+    std::set<std::string> readers;
+    for( const std::pair<npc *, std::string> &elem : fun_learners ) {
```
```suggestion
    for( const std::pair<npc *const, std::string> &elem : fun_learners ) {
```
Ditto.

actual-nh reviewed (commented):
I am not great at C++ (I can manage C modifications; I prefer Python or Perl for new programs); this is from looking at the Clang 10 errors.

NickGrease commented:
While the new menu order might seem more 'natural', I'm guessing 'read until...' is used a vast majority of the time, making it reasonable to either be the first entry, or selected by default - otherwise you're simply introducing additional keystrokes. /2cents

Saicchi commented:
I changed it to select the second option by default

It seems there is a really annoying bug when deserializing an `item_location` with type `in_container`, it can't find the player with the id that has the item, but things work fine when the `item_location` type is `character`.  This is really annoying and I don't have any way to fix, it works fine when the item is wielded but if its in a container, it gets an error when getting the `parent` `item_location`.

BrettDong commented on src/activity_actor.cpp:
```diff
@@ -2597,6 +2603,487 @@ std::unique_ptr<activity_actor> milk_activity_actor::deserialize( JsonIn &jsin )
+        you.has_identified( book->typeId() ) ) {
+        const SkillLevel &skill_level = you.get_skill_level_object( skill );
+        //~ skill_name current_skill_level -> next_skill_level (% to next level)
+        return string_format( pgettext( "reading progress", "%s %d -> %d (%d%%)" ),
```
```suggestion
        return string_format( pgettext( "reading progress", "%1$s %2$d -> %3$d (%4$d%%)" ),
```

BrettDong commented on src/npc.cpp:
```diff
@@ -922,146 +923,49 @@ int npc::time_to_read( const item &book, const player &reader ) const
-            start_read( book, pl );
+    book = book.obtain( *npc_character );
+    if( can_read( *book, fail_reasons ) ) {
+        if( get_player_view().sees( pos() ) ) {
+            add_msg( m_info, _( "%s starts reading." ), disp_name() );
```
```suggestion
        add_msg_if_player_sees( pos(), _( "%s starts reading." ), disp_name() );
```

BrettDong commented on src/npc.cpp:
```diff
@@ -945,8 +945,7 @@ void npc::do_npc_read()
     if( can_read( *book, fail_reasons ) ) {
-        if( get_player_view().sees( pos() ) ) {
-            add_msg( m_info, _( "%s starts reading." ), disp_name() );
+        add_msg_if_player_sees( pos(), _( "%s starts reading." ), disp_name() );
```
Actually should have removed the curly bracket below as well but I am unable to include that in my comment

BrettDong commented on src/activity_actor.cpp:
```diff
@@ -2597,6 +2603,487 @@ std::unique_ptr<activity_actor> milk_activity_actor::deserialize( JsonIn &jsin )
+            ( who.is_npc() || query_yn( _( "Wield the %s and start reading?" ), book->tname() ) ) )   {
+
+            if( !who.wield( *book ) ) {
+                add_msg( m_bad, "%s", who.can_wield( *book ).c_str() );
```
```suggestion
                add_msg( m_bad, who.can_wield( *book ).c_str() );
```

BrettDong commented on src/activity_actor.cpp:
```diff
@@ -2597,6 +2603,487 @@ std::unique_ptr<activity_actor> milk_activity_actor::deserialize( JsonIn &jsin )
+    min_ex = learner.adjust_for_focus( min_ex );
+    max_ex = learner.adjust_for_focus( max_ex );
+
+    if( max_ex < 2 ) {
+        max_ex = 2;
+    }
+
+    if( max_ex > 10 ) {
+        max_ex = 10;
+    }
+
+    if( max_ex < min_ex ) {
+        max_ex = min_ex;
+    }
```
Can be simplified with
```suggestion
    max_ex = clamp( max_ex, 2, 10 );
    max_ex = std::max( min_ex, max_ex );
```

Saicchi commented:
Thanks @BrettDong , keep sending more changes and I'll implement them while doing a rebase.

BrettDong commented on src/activity_actor.cpp:
```diff
@@ -2597,6 +2603,478 @@ std::unique_ptr<activity_actor> milk_activity_actor::deserialize( JsonIn &jsin )
+
+    int difficulty = std::max( 1, style_to_learn->learn_difficulty );
+    difficulty = std::max( 1, 20 + difficulty * 2 - you.get_skill_level( skill_used ) * 2 );
+    add_msg_debug( _( "Chance to learn one in: %d" ), difficulty );
```
Debug message does not need to be translated.
```suggestion
    add_msg_debug( "Chance to learn one in: %d", difficulty );
```

BrettDong commented on src/npc.cpp:
```diff
@@ -922,146 +923,48 @@ int npc::time_to_read( const item &book, const player &reader ) const
+    book = book.obtain( *npc_character );
+    if( can_read( *book, fail_reasons ) ) {
+        add_msg_if_player_sees( pos(), _( "%s starts reading." ), disp_name() );
         }
```
```suggestion
```

BrettDong commented:
Finished reviewing the code, now I am testing it.

Saicchi commented:
I took at look at `ACT_GUNMOD_ADD` to figure out how to avoid the `item_location` loading problem. Now you or npcs don't need to wield the book anymore and it works more like before.

BrettDong commented:
Seems like you've included some other irrelevant commits here..?

Saicchi commented:
Where? I did a rebase and a small fix.

BrettDong commented:
I think the `Update melee_dodge_hit_test.cpp` commits belong to some other PRs.

Saicchi commented:
Oops you're right. That must have happend while doing a rebase somewhere somehow.

BrettDong commented on src/activity_actor.cpp:
```diff
@@ -2686,6 +2692,473 @@ std::unique_ptr<activity_actor> disassemble_activity_actor::deserialize( JsonIn
+
+void read_activity_actor::do_turn( player_activity &act, Character &who )
+{
+    if( who.fine_detail_vision_mod() > 4 ) {
+        //It got too dark during the process of reading, bail out.
+        act.set_to_null();
+        who.add_msg_if_player( m_bad, _( "It's too dark to read!" ) );
+        return;
+    }
+
+    if( !bktype.has_value() ) {
+        bktype = book->type->use_methods.count( "MA_MANUAL" ) ?
+                 book_type::martial_art : book_type::normal;
+    }
+
+    if( who.is_avatar() ) {
```
Needs to backport #46615 and solve merge conflict 😂
```suggestion
    if( !bktype.has_value() ) {
        bktype = book->type->use_methods.count( "MA_MANUAL" ) ?
                 book_type::martial_art : book_type::normal;
    }

    if( who.is_avatar() ) {
        // the following check doesn't work for NPCs because it is counted for player's submap and z-level only
        if( who.fine_detail_vision_mod() > 4 ) {
            //It got too dark during the process of reading, bail out.
            act.set_to_null();
            who.add_msg_if_player( m_bad, _( "It's too dark to read!" ) );
            return;
        }
```

BrettDong reviewed (approved):
Tested reading various books, with NPCs, with interruptions. Haven't observed any anomaly.

PatrikLundell commented:
Some accepted PR has changed a file resulting in a conflict. Adding comment as Github doesn't provide you with any notification.

Saicchi commented:
I'll postpone doing a rebase to fix conflicts until 0.F or until someone asks me to.

Saicchi commented:
I've rebased it to a more recent version of experimental ( from a few days ago ), I've done some basic testing and it seems to work. When rebasing I got some merge conflicts so please tell me if I forgot any change that happend in the past 6 months regarding book reading.

actual-nh commented:
Any progress on the merge conflicts?

Saicchi commented:
I got them all fixed but in the process I might have reverted a change accidentaly. Right now I should rebase it again for the new experimental.

kevingranade commented:
Please review my merge, there's a lot of moving pieces in here.

Saicchi commented:
I've rebased this PR to `cdda-experimental-2021-07-08-1134` and fixed an issue where the activity wasn't restarted properly when canceled.

I-am-Erk commented:
More merge conflicts, because we're merging things at an unbelievable rate

kevingranade commented:
Need a careful look at what changed in do_read in master to make sure we don't overwrite changes to it.

Night-Pryanik commented:
@Saicchi could you please resolve conflicts?

Saicchi commented:
I've rebased it for experimental build 2021-07-17-0839 and added changes in `avatar::do_read` and `npc::finish_read` from #47656. With some testing it should be the same as before the changes, but training with a martial arts book seems to be taking less stamina. This has been rebased so many times that I'm worried I've missed something.

I-am-Erk commented:
I am going to merge this now then, I think it will be easier to fix problems that may arise than to keep rebasing this.
